### PR TITLE
simple preheat for storage client in graph service

### PR DESCRIPTION
**Pull request by critical27**
Since we have many poc are going online, one problem is that first query of every partition probably would fail because of StorageClient has no leader info.

This PR is a simple version of preheat after #1236 and #1548 merged, when MetaClient init, it would try to get leader distribution from meta server, and storage client in graph could use it directly.

For simplicity, I only load leader distribution only once instead of update it periodically.  Which bring a problem, **you have to start meta and storage service first, after a few seconds, then you can start graph services**. 

PS: Just like problem above, after `balance leader`, the leader info probably is wrong. If you want to preheat again, you need to restart graph.

@Amber1990Zhang, in doc, we could advice user start meta and storage first, and start graph later.

### CODE REVIEW

**Comment by dangleptr** [comment]
>>> Which bring a problem, you have to start meta and storage service first, after a few seconds, then you can start graph services.

We could handle the case by lazy init the leadersMap when accessing it.

**Comment by dangleptr** [comment]
>>> PS: Just like problem above, after balance leader, the leader info probably is wrong. If you want to preheat again, you need to restart graph.

I think we don't need to repreheat after the leader balanced.

**Comment by critical27** [comment]
> > > > Which bring a problem, you have to start meta and storage service first, after a few seconds, then you can start graph services.
> 
> We could handle the case by lazy init the leadersMap when accessing it.

Good suggestion

File: src/storage/client/StorageClient.cpp [hunk 1]
```diff
@@ -502,6 +567,11 @@ folly::SemiFuture<StorageRpcResponse<storage::cpp2::GeneralResponse>>
 StorageClient::get(GraphSpaceID space,
                    const std::vector<std::string>& keys,
                    folly::EventBase* evb) {
+    if (!loadLeaderBefore_ && !loadLeader()) {
```
Comment (dangleptr) [inline src/storage/client/StorageClient.cpp]: Could we move loadLeader inside this->leader() function.

Comment (critical27) [reply src/storage/client/StorageClient.cpp]: fixed

**Comment by dangleptr** [comment]
> Please note that: Since there involves synchronous RPC calls, there might be performance issues when bursting requests arrives(especially async ones) if `loadLeader` has not been invoked.

Yes, we should handle this problem.

**Comment by critical27** [comment]
> > Please note that: Since there involves synchronous RPC calls, there might be performance issues when bursting requests arrives(especially async ones) if `loadLeader` has not been invoked.
> 
> Yes, we should handle this problem.

I've considered this problem before, how about double check the loadLeaderBefore_ in loadLeader, WDYT?

File: src/meta/client/MetaClient.cpp [hunk 2]
```diff
@@ -1779,5 +1779,38 @@ Status MetaClient::refreshCache() {
...
+
+    LeaderMap leaderMap;
+    auto hostItems = std::move(ret).value();
+    for (auto& item : hostItems) {
+        auto hostAddr = HostAddr(item.hostAddr.ip, item.hostAddr.port);
+        for (auto& spaceEntry : item.get_leader_parts()) {
+            auto spaceName = spaceEntry.first;
+            auto status = getSpaceIdByNameFromCache(spaceName);
+            if (!status.ok()) {
+                continue;
```
Comment (dangleptr) [inline src/meta/client/MetaClient.cpp]: Why continue here?

Comment (critical27) [reply src/meta/client/MetaClient.cpp]: Cause `show hosts` return map of <space name, leader>, we have to convert to spaceId here. We just can ignore the if the space exists in `show hosts` but not in `show spaces`, actually they are both from meta, the `continue` rarely can hit.

Comment (dangleptr) [reply src/meta/client/MetaClient.cpp]: Fine.

**Comment by jude-zhu** [comment]
relate to #735

**Pull request merged**